csrc/optimization/layout_inference.cpp

```diff
@@ -0,0 +1,236 @@
+//    |    alloc dom [i0, i2, i1]
+//    |
+//    |
+//    BroadcastOp
+//    |
+//    v
+//   TV1 rfactor dom [i0, b3, i1, i2, b4]
+//         alloc dom [b3, b4, i0, i2, i1]  *see note 1
+//                     1,  4,
+//                             0,  3,  2   *see note 2
+//   note 1:
+//       keeping the alloc domain from input [i0, i2, i1]
+//       stack all broadcast in rfactor [b3, b4].
+//       concat([b3, b4], [i0, i2, i1])
+//
+//   note 2:
+//       computing the new output memory format
+//       We'll scan through the rfactor domain of output where:
+//       a. insert any broadcast iterdomain index as we encounter them
+//       b. adjust the index of entry from input's rfactor domain
+//
+//   so output TV1 will have stride order {1, 4, 0, 3, 2}
+void MemoryFormatInferencer::handle(const BroadcastOp* op) {
+  TensorView* out = dynamic_cast<TensorView*>(op->out());
+  if (out == nullptr) {
+    return;
+  }
+  TensorView* in = op->in()->as<TensorView>();
+  if (const auto& iter = format_map_.find(in); iter != format_map_.end()) {
+    MemoryFormat out_format;
+    int64_t out_rank = static_cast<int64_t>(out->nDims());
+
+    int broadcast_seen_so_far = 0;
+    std::vector<int64_t> offset_table(in->nDims(), 0);
+    int offset_entry = 0;
+
+    for (auto i : c10::irange(out_rank)) {
+      if (op->isBroadcastDim(i)) {
+        broadcast_seen_so_far++;
+        // broadcast dimensions are default to outer dimensions
+        // see note 2.a
+        out_format.push_back(i);
+      } else {
+        // adjusting entry point by recording index compensation
+        // i.e. broadcast dimensions inserted on the left of the old iterdomain
+        // see note 2.b
+        offset_table[offset_entry++] = broadcast_seen_so_far;
+      }
+    }
+
+    for (auto i : c10::irange(in->nDims())) {
+      auto format_entry = iter->second[i];
+      out_format.push_back(format_entry + offset_table[format_entry]);
+    }
+    format_map_[out] = out_format;
+  }
+}
+
+} // namespace
+
+// Note [ Memory Format Propagation ]
+//
+// The propagation tries to propagate memory format from inputs to the entire
+// fusion:
+//   1. Iterates through all inputs, looking for TensorView with allocatoin
+//   domain that's a permutation of its corresponding rfactor domain and record
+//   it as the memory format of the tensor;
+//   2. Traverse the fusion IR, propagate memory format and record results in
+//   memory_format_map.
+std::unordered_map<const TensorView*, MemoryFormat> inferenceMemoryFormat(
+    Fusion* fusion) {
+  std::unordered_map<const TensorView*, MemoryFormat> memory_format_map;
+
+  // Note: we only consider simple permutation of allocation domain to rfactor
+  // domain.
+  for (auto tv : ir_utils::filterByType<TensorView>(fusion->inputs())) {
+    std::optional<MemoryFormat> permutation = ir_utils::computePermutation(
+        TensorDomain::noReductions(tv->getMaybeRFactorDomain()),
+        tv->getMaybeAllocationDomain());
```

Comment:
Should this be `TensorDomain::noReductions(tv->getMaybeAllocationDomain())`? IIRC allocation domain do have these reductions, although it makes no sense to do so.

Reply:
Also, should we make sure that reductions in the allocation domain are correctly handled?